cclunit-framework-source/doc/CCLUTMOCKING.md

```diff
@@ -0,0 +1,268 @@
+  
+@param tableName  
+&nbsp;&nbsp;&nbsp;&nbsp;The table to be mocked.  
+  
+Example:  
+call cclutCreateMockTable("person")
+
+**cclutRemoveMockTable(tableName = vc)**
+
+Removes the mock table.  If the table was already created, it will also be dropped.  If the table is not currently  
+mocked, it will return silently.  tableName is required.  
+  
+@param tableName  
+&nbsp;&nbsp;&nbsp;&nbsp;The table that is mocked.  
+  
+Example:  
+call cclutRemoveMockTable("person")
+
+**cclutRemoveAllMockTables(null)**
+
+Removes all mock tables.  Any tables that have already been created will also be dropped.  
+  
+Example:  
+call cclutRemoveAllMockTables(null)
+
+**cclutAddMockData(tableName = vc, rowData = vc)**
+
+Add a row of mock data to a table.  tableName and rowData are required.  tableName must have already been created through cclutCreateMockTable() or an error will be thrown.  rowData is a pipe-delimited string for each column in the same order that was used in cclutDefineMockTable().  For character fields, the backslash (\\) will serve as an escape character.  For date fields, the value in rowData will be supplied to the cnvtdatetime() function.  All other values will be passed as-is.  
+  
+Supported escape values  
+\\| = | (to represent a pipe in character fields)  
+\\\\ = \ (to represent a backslash in character fields)  
+\null = null (no value will be inserted into the column)  
+  
+@param tableName  
+&nbsp;&nbsp;&nbsp;&nbsp;The table to which the data will be added.  
+@param rowData  
+&nbsp;&nbsp;&nbsp;&nbsp;A pipe-delimited string of data to be inserted into the table.  
+  
+Example:  
+call cclutDefineMockTable("person", "person_id|name_last|name_first|birth_dt_tm", "f8|vc|vc|dq8")  
+call cclutCreateMockTable("person")  
+call cclutAddMockData("person", "1.0|Washington|George|01-JAN-1970 00:00") ;Will add George Washington  
+call cclutAddMockData("person", "2.0|A\\|d\\\\ams|John|02-FEB-1971 11:11") ;Will add John A|d\ams  
+call cclutAddMockData("person", "3.0|Jefferson|\null|03-MAR-1972 22:22") ;Will add Jefferson (no first name)  
+call cclutAddMockData("person", "4.0|Madison||04-APR-1973 10:33") ;Will add Madison (empty string for first name)
+
+**cclutClearMockData(tableName = vc)**
+
+Clears all data from the mock table.  This is functionally similar to a truncate.  tableName is required.  The table  
+must have been created through cclutCreateMockTable() or else an error will be thrown.  
+  
+@param tableName  
+&nbsp;&nbsp;&nbsp;&nbsp;The table to be cleared.  
+  
+Example:  
+call cclutClearMockData("person")
+
+**cclutAddMockImplementation(originalName = vc, replaceName = vc)**
+
+Adds a mock implementation to be utilized by cclutExecuteProgramWithMocks.  This can include record structures, subroutines, or other programs.  originalName and replaceName are required.  If originalName is already being mocked, the mock will be overwritten with the new replaceName.  
```

Comment:
> Reading "cclutExecuteProgramWithMocks" at this point causes quite a jolt because this is the very first mention of "cclutExecuteProgramWithMocks". Rewording this current description to not mention "cclutExecuteProgramWithMocks" would leave it sounding wishy-washy. On the other hand rewording the doc for cclutExecuteProgramWithMocks to be somewhat generic and moving it to the beginning would resolve the discontinuity. Besides, it makes total sense to open with a discussion of cclutExecuteProgramWithMocks because it is the showcase function. All others exist only for its support. See the suggested changes for cclutExecuteProgramWithMocks below.

I'll keep the cclutExecuteProgramWithMocks discussion in the thread for that function so as not to be confusing on the review, but I agree with your assessment.

> Regarding the name of this subroutine....
"Add" seems like the wrong verb. "Define", "Specify", "Set" ?
"Implementation" seems like the wrong noun. "Entity", "Item?
"Implementation" makes sense for uars, subroutines and programs but not really for variables or record structures.
The term "object" is an option and it gets used later in the description, but (1) some types of mocked items don't feel like objects and (2) CCL is trying to become object-oriented giving the word "object" a rather specific meaning.

For this subroutine, I'd be fine with cclutDefineMockEntity.  The "Define" keeps it consistent with what we're using for the tables.

> This description should also mention uars.

Will do.

> Better wording?
"If a mock has already been specified for the originalName, the new replacementName will be used instead". (or perhaps "the new replacementName will override"?)

I like the first version, since it's not overriding so much as it's overwriting (unless you prefer "overwrite" to "used").

> Including "Name" in the parameter names feels odd. In addition, using the terms "name" and "replacement" feels like an implementation leak.
How about simply naming them "original" and "mock" ?

I'm fine with using "mock" instead of "replace", but the "name" convention is used in other parts of the mocking framework as well (such as tableName and programName).  I kind of like that consistency, and it's also clear that the "name" (as a vc string) should be sent in, and not, for example, the record structure itself.  Given that it's possible to send record structures to subroutines, someone might mistakenly think that the record structure itself is what is needed instead of its name.  Do you think that would be a valid concern?